spec/history.cddl

```diff
@@ -0,0 +1,34 @@
+canister-change-entry = tagged<{
+  timestamp : timestamp
+  origin : canister-change-origin
+  change : canister-change
+}>
+
+canister-change-origin = {
+  from : "user"
+  user_id : principal
+  //
+  from : "canister"
+  canister_id : principal
+  canister_history_index : unsigned
+  //
+  from : "system"
+}
+
+canister-change = {
+  kind : "canister-creation" / "code-uninstall"
+  //
+  kind : "code-install" / "code-reinstall" / "code-upgrade"
```

Comment:
shouldn't the enum match the `install_code` spec?

https://github.com/dfinity/interface-spec/blob/f29fb083893c4d19de6373e1da6f74203a28679f/spec/ic.did#L93

Reply:
I don't see this as a strict requirement since there also exist canister change kinds that are not members of this variant, e.g., `canister-creation` or `code-uninstall`. Do you see a reason why these three modes need to be encoded literally?

Reply:
Not a did expert but assuming the two informations are linked, if I would implement this in object base language, the kind would probably be an extension of the spec because otherwise the types would have to be mapped. As the map would become an important part of the conversion and tracability of the information, I would rather like to spare it.

But then indeed, if the `canister-change { kind }` and `install_code { mode }` are not related at all, then it can be different.

Being said, I am not a scientist or backend architect, so don't mind me if you rather like to use a custom variant.